### PR TITLE
Upgrade cuda drivers from 12.1 to 12.4

### PR DESCRIPTION
**Pull request by atheo89**
Related to: https://issues.redhat.com/browse/RHOAIENG-13129 
## How Has This Been Tested?

https://issues.redhat.com/browse/RHOAIENG-13451 

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: cuda/ubi9-python-3.11/Dockerfile [hunk 1]
```diff
@@ -12,26 +12,26 @@ LABEL name="odh-notebook-cuda-ubi9-python-3.11" \
...
+ENV NV_CUDA_CUDART_VERSION 12.4.127-1
 
-COPY cuda.repo-x86_64 /etc/yum.repos.d/cuda.repo
+COPY cuda.repo-arm64 /etc/yum.repos.d/cuda.repo
```
Comment (jiridanek) [inline cuda/ubi9-python-3.11/Dockerfile:23]: Isn't this a mistake, should be `amd64`, not `arm64`?

File: cuda/ubi9-python-3.11/Dockerfile [hunk 2]
```diff
@@ -12,26 +12,26 @@ LABEL name="odh-notebook-cuda-ubi9-python-3.11" \
...
+ENV NVIDIA_REQUIRE_CUDA "cuda>=12.4 brand=tesla,driver>=470,driver<471 brand=unknown,driver>=470,driver<471 brand=nvidia,driver>=470,driver<471 brand=nvidiartx,driver>=470,driver<471 brand=geforce,driver>=470,driver<471 brand=geforcertx,driver>=470,driver<471 brand=quadro,driver>=470,driver<471 brand=quadrortx,driver>=470,driver<471 brand=titan,driver>=470,driver<471 brand=titanrtx,driver>=470,driver<471 brand=tesla,driver>=525,driver<526 brand=unknown,driver>=525,driver<526 brand=nvidia,driver>=525,driver<526 brand=nvidiartx,driver>=525,driver<526 brand=geforce,driver>=525,driver<526 brand=geforcertx,driver>=525,driver<526 brand=quadro,driver>=525,driver<526 brand=quadrortx,driver>=525,driver<526 brand=titan,driver>=525,driver<526 brand=titanrtx,driver>=525,driver<526 brand=tesla,driver>=535,driver<536 brand=unknown,driver>=535,driver<536 brand=nvidia,driver>=535,driver<536 brand=nvidiartx,driver>=535,driver<536 brand=geforce,driver>=535,driver<536 brand=geforcertx,driver>=535,driver<536 brand=quadro,driver>=535,driver<536 brand=quadrortx,driver>=535,driver<536 brand=titan,driver>=535,driver<536 brand=titanrtx,driver>=535,driver<536"
+ENV NV_CUDA_CUDART_VERSION 12.4.127-1
```
Comment (jiridanek) [inline cuda/ubi9-python-3.11/Dockerfile:21]: How can we have CUDA 12.4 when the latest version mentioned by Tensorflow is 12.3?


Version | Python version | Compiler | Build tools | cuDNN | CUDA
-- | -- | -- | -- | -- | --
tensorflow-2.17.0 | 3.9-3.12 | Clang 17.0.6 | Bazel 6.5.0 | 8.9 | 12.3



https://www.tensorflow.org/install/source#gpu

File: cuda/ubi9-python-3.11/Dockerfile [hunk 3]
```diff
@@ -3,7 +3,7 @@ FROM ${BASE_IMAGE}
 
 LABEL name="odh-notebook-cuda-ubi9-python-3.11" \
       summary="CUDA Python 3.11 base image for ODH notebooks" \
-      description="CUDA Python 3.11 builder image based on UBI8 for ODH notebooks" \
+      description="CUDA Python 3.11 builder image based on UBI9 for ODH notebooks" \
```
Comment (jiridanek) [inline cuda/ubi9-python-3.11/Dockerfile:6]: 👍🏽

File: cuda/ubi9-python-3.11/Dockerfile (hunk 2 above)
Comment (atheo89) [reply cuda/ubi9-python-3.11/Dockerfile:21]: Yes this needs to be verified open ticket here. https://issues.redhat.com/browse/RHOAIENG-13451 

Just FWI till previous week on this gpu matrix was till here
Version | Python version | Compiler | Build tools | cuDNN | CUDA
-- | -- | -- | -- | -- | --
tensorflow-2.16.1	| 3.9-3.12 | Clang 17.0.6 | Bazel 6.5.0 | 8.9 | 12.3

File: cuda/ubi9-python-3.11/Dockerfile (hunk 1 above)
Comment (atheo89) [reply cuda/ubi9-python-3.11/Dockerfile:23]: Its not mistake, this is how it is been defined on the nvidia [dockerfile](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/12.4.1/ubi9/base/Dockerfile#L17 )

File: cuda/ubi9-python-3.11/Dockerfile (hunk 2 above)
Comment (atheo89) [reply cuda/ubi9-python-3.11/Dockerfile:21]: Open PR for 12.3 is here https://github.com/opendatahub-io/notebooks/pull/711

File: cuda/ubi9-python-3.11/Dockerfile (hunk 1 above)
Comment (jiridanek) [reply cuda/ubi9-python-3.11/Dockerfile:23]: in that case nvidia has it wrong

Comment (atheo89) [reply cuda/ubi9-python-3.11/Dockerfile:23]: probably, because also the build failed apart

Comment (jstourac) [reply cuda/ubi9-python-3.11/Dockerfile:23]: Please, note that there are two FROM base sections in that linked dockerfile - first one is amd64 arch and the second one is the arm-64 arch.

If I understand it correctly the way they have implemented it, they share this dockerfile to build both platforms by specifying the `TARGETARCH` variable. So we should simply choose the arch we are interested in only - amd64.

Comment (atheo89) [reply cuda/ubi9-python-3.11/Dockerfile]: Right that was the issue, I changed rpm

**Comment by atheo89** [comment]
Based on the testing that @daniellutz did yesterday seems that we are good to go with 12.4.
Please take a final look in order to have this in place.

File: cuda/ubi9-python-3.11/Dockerfile [hunk 4]
```diff
@@ -12,26 +12,26 @@ LABEL name="odh-notebook-cuda-ubi9-python-3.11" \
...
 # Install CUDA base from:
-# https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/12.1.1/ubi9/base/Dockerfile
+# https://gitlab.com/nvidia/container-images/cuda/-/tree/master/dist/12.4.1/ubi9/base
 USER 0
 WORKDIR /opt/app-root/bin
 
-ENV NVARCH x86_64
-ENV NVIDIA_REQUIRE_CUDA "cuda>=12.1 brand=tesla,driver>=470,driver<471 brand=unknown,driver>=470,driver<471 brand=nvidia,driver>=470,driver<471 brand=nvidiartx,driver>=470,driver<471 brand=geforce,driver>=470,driver<471 brand=geforcertx,driver>=470,driver<471 brand=quadro,driver>=470,driver<471 brand=quadrortx,driver>=470,driver<471 brand=titan,driver>=470,driver<471 brand=titanrtx,driver>=470,driver<471 brand=tesla,driver>=525,driver<526 brand=unknown,driver>=525,driver<526 brand=nvidia,driver>=525,driver<526 brand=nvidiartx,driver>=525,driver<526 brand=geforce,driver>=525,driver<526 brand=geforcertx,driver>=525,driver<526 brand=quadro,driver>=525,driver<526 brand=quadrortx,driver>=525,driver<526 brand=titan,driver>=525,driver<526 brand=titanrtx,driver>=525,driver<526"
-ENV NV_CUDA_CUDART_VERSION 12.1.105-1
+ENV NVARCH sbsa
```
Comment (jstourac) [inline cuda/ubi9-python-3.11/Dockerfile:19]:
```suggestion
ENV NVARCH x86_64
```

**Comment by jstourac** [comment]
One comment, otherwise LGTM.

File: cuda/ubi9-python-3.11/Dockerfile (hunk 4 above)
Comment (atheo89) [reply cuda/ubi9-python-3.11/Dockerfile:19]: tnx for noticing that, i had this change on my repo but i never push it :facepalm: Opened a new PR here: https://github.com/opendatahub-io/notebooks/pull/717

**Pull request merged**